lib/ex_aws/sns.ex

```diff
@@ -169,6 +169,77 @@ defmodule ExAws.SNS do
     request("ListSubscriptions", %{"NextToken" => next_token})
   end
 
+  ## Messages
+  ######################
+
+  @notification_params ["Message", "MessageId", "Subject", "Timestamp", "TopicArn", "Type"]
+  @optional_notification_params ["Subject"]
+  @confirmation_params ["Message", "MessageId", "SubscribeURL", "Timestamp", "Token", "TopicArn", "Type"]
+  @signature_params ["SignatureVersion", "Signature", "SigningCertURL"]
+  @message_types ["SubscriptionConfirmation", "UnsubscribeConfirmation", "Notification"]
+
+  @doc "Verify message signature"
+  def verify_message(message_params) do
+    with :ok <- validate_message_params(message_params),
+      :ok <- validate_signature_version(message_params["SignatureVersion"]),
+      {:ok, public_key} <- ExAws.SNS.PublicKeyCache.get(message_params["SigningCertURL"]) do
+        message_params
+        |> get_string_to_sign
+        |> verify(message_params["Signature"], public_key)
+    else
+      {:error, error} -> {false, error}
+    end
+  end
+
+  defp validate_message_params(message_params) do
+    with {:ok, required_params} <- get_required_params(message_params["Type"]) do
+      case required_params -- Map.keys(message_params) do
+        [] -> :ok
+        missing_params -> {:error, "The following parameters are missing: #{inspect missing_params}"}
+      end        
+    end
+  end
+
+  defp get_required_params(message_type) do
+    case message_type do
+      "Notification" -> {:ok, (@notification_params -- @optional_notification_params) ++ @signature_params}
+      "SubscriptionConfirmation" -> {:ok, @confirmation_params ++ @signature_params}
+      "UnsubscribeConfirmation" -> {:ok, @confirmation_params ++ @signature_params}
+      type when is_binary(type) -> {:error, "Invalid Type, expected one of #{inspect @message_types}"}
+      type when is_nil(type) -> {:error, "Missing Type c (Type)"}
+      type -> {:error, "Invalid Type's type, expected a String, got #{inspect type}"}
+    end
+  end
+
+  defp validate_signature_version(version) do
+    case version do
+      "1" -> :ok
+      sig when is_binary(sig) -> {:error, "Unsupported SignatureVersion, expected \"1\", got #{version}"}
+      sig -> {:error, "Invalid SignatureVersion format, expected a String, got #{version}"}
+    end
+  end
+
+  defp get_string_to_sign(message_params) do
+    message_params
+    |> Map.take(get_params_to_sign(message_params["Type"]))
+    |> Enum.reduce("", fn {key, value}, acc -> acc <> key <> "\n" <> value <> "\n" end)
```

Comment:
Sorry I wasn't super clear. What I had in mind was something more like
```
|> Enum.map(fn {key, value} -> [to_string(key), "\n", to_string(value), "\n"] end)
|> IO.iodata_to_binary
```

It's more efficient to build up IOData in this case than concatenating binaries. Doesn't make a huge difference but it's also pretty simple code. I almost always prefer `{key, value}` over `elem` btw, just seems clearer.